### ISSUE TITLE
Event auditing

### ISSUE THREAD

**Issue by Firstyear** [kanidm/kanidm#22]
In some cases we want to record audit events on objects for transparency. For example, on failed authentication we want to write to the object about the failed authentication. We want to audit account impersonations, password changes, etc. 

For some events, these should imply a write to the object itself that allows the object to self-read it's audit information about what has happened and who is using the account.

**Comment by yaleman** [comment]
The more I think about the transport/pipeline for these is that OpenTelemetry's model is one of the best options - Kanidm publishes the events into a pipeline and then if folks want to dump them to disk/stdout then that's doable, or directly publish to one of the many SIEM/logging providers, or a database etc.

Example here: https://github.com/open-telemetry/opentelemetry-rust/blob/18f6c01ee35f2e7909fe46a815ecd6f958ecef71/examples/logs-basic/src/main.rs

**Comment by Firstyear** [comment]
Yeah, I think for now internally to Kanidm each server will have it's own queue (quwu) then I think opentelemetry can be a backend to that. I wouldn't want opentelemetry in the internal server paths as I don't know much about it's internals/performance, so better to control the queueing internally then we can have that feed opentelemetry in a seperate thread/worker.

**Comment by yaleman** [comment]
Yep, as I understand it, logs are just hooked into the default log provider and it sits off the side doing its thing.

**Comment by Firstyear** [comment]
> Yep, as I understand it, logs are just hooked into the default log provider and it sits off the side doing its thing.

Yep, and later we can have that task funnel events into something else like opentelemetry etc.

**Comment by dvv** [comment]
Hi!

I wonder what's the status of the progress?
Can it be used to catch details of changes made to person/group/oauth2 via API or client tool?
If it's not at all intended for the above how it's possible to provide kinda eventsource of changes made to kanidm objects?

TIA

**Comment by yaleman** [comment]
At the moment there's no design or implementation.
